frame/support/src/traits/tokens/fungibles/hold.rs

```diff
@@ -0,0 +1,455 @@
+		Restriction::{self, Free, OnHold},
+	},
+};
+use scale_info::TypeInfo;
+use sp_arithmetic::{
+	traits::{CheckedAdd, CheckedSub, Zero},
+	ArithmeticError,
+};
+use sp_runtime::{DispatchError, DispatchResult, Saturating, TokenError};
+
+use super::*;
+
+/// Trait for inspecting a fungible asset whose accounts support partitioning and slashing.
+pub trait Inspect<AccountId>: super::Inspect<AccountId> {
+	/// An identifier for a hold. Used for disambiguating different holds so that
+	/// they can be individually replaced or removed and funds from one hold don't accidentally
+	/// become unreserved or slashed for another.
+	type Reason: codec::Encode + TypeInfo + 'static;
+
+	/// Amount of funds on hold (for all hold reasons) of `who`.
+	fn total_balance_on_hold(asset: Self::AssetId, who: &AccountId) -> Self::Balance;
+
+	/// Get the maximum amount that the `total_balance_on_hold` of `who` can be reduced successfully
+	/// based on whether we are willing to force the reduction and potentially go below user-level
+	/// restrictions on the minimum amount of the account. Note: This cannot bring the account into
+	/// an inconsistent state with regards any required existential deposit.
+	///
+	/// Always less than `total_balance_on_hold()`.
+	fn reducible_total_balance_on_hold(
+		asset: Self::AssetId,
+		who: &AccountId,
+		force: Fortitude,
+	) -> Self::Balance;
+
+	/// Amount of funds on hold (for the given reason) of `who`.
+	fn balance_on_hold(
+		asset: Self::AssetId,
+		reason: &Self::Reason,
+		who: &AccountId,
+	) -> Self::Balance;
+
+	/// Returns `true` if it's possible to place (additional) funds under a hold of a given
+	/// `reason`. This may fail if the account has exhausted a limited number of concurrent
+	/// holds or if it cannot be made to exist (e.g. there is no provider reference).
+	///
+	/// NOTE: This does not take into account changes which could be made to the account of `who`
+	/// (such as removing a provider reference) after this call is made. Any usage of this should
+	/// therefore ensure the account is already in the appropriate state prior to calling it.
+	fn hold_available(asset: Self::AssetId, reason: &Self::Reason, who: &AccountId) -> bool;
+
+	/// Check to see if some `amount` of funds of `who` may be placed on hold with the given
+	/// `reason`. Reasons why this may not be true:
+	///
+	/// - The implementor supports only a limited number of concurrernt holds on an account which is
+	///   the possible values of `reason`;
+	/// - The total balance of the account is less than `amount`;
+	/// - Removing `amount` from the total balance would kill the account and remove the only
+	///   provider reference.
+	///
+	/// Note: we pass `true` as the third argument to `reducible_balance` since we assume that if
+	/// needed the balance can slashed. If we are using a simple non-forcing reserve-transfer, then
+	/// we really ought to check that we are not reducing the funds below the freeze-limit (if any).
```

Comment:
```suggestion
	/// Note: we pass `Fortitude::Force` as the last argument to `reducible_balance` since we assume that if
	/// needed the balance can slashed. If we are using a simple non-forcing reserve-transfer, then
	/// we really ought to check that we are not reducing the funds below the freeze-limit (if any).
```